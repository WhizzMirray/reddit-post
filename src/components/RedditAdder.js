import React from 'react';
import { connect } from 'react-redux';
import {fetchPostsIfNeeded} from '../actions/redditActions';
import {openSettingsModal} from '../actions/utilActions';
// \A[A-Za-z0-9][A-Za-z0-9_]{1,20}\Z
const regex = new RegExp('^[A-Za-z0-9][A-Za-z0-9_]{1,20}$');

const ErrorMessage = ({message,closeError}) => (
  <div className="alert alert-danger" role="alert" style={{marginBottom:'0px',padding:'8px 15px'}}>
    <span className="glyphicon glyphicon-exclamation-sign" aria-hidden="true" style={{paddingRight:'5px'}}></span>
    <span className="sr-only">Error:</span>
    <span className="pull-right clickable">
      <button className="error-close btn btn-default" type="button" onClick={closeError}>
        <span className="glyphicon glyphicon-remove"/>
      </button>
    </span>
    {message}
  </div>
);

class redditAdder extends React.Component {
  constructor(props){
    super(props);
    this.state={
      value : "",
      error : ""
    }
  }
  render(){

    return(
      <form onSubmit={(e)=>this.handleSubmit(e)}>
        <div className="input-group">
            <input type="text" className="form-control" placeholder="Type the subreddit you want to add ex: 'worldnews'"
              onChange={this.handleChange.bind(this)} value={this.state.value}/>
            <span className="input-group-btn">
              <button className="btn btn-default" type="submit" style={{border:'1px solid #cccccc'}}>
                <span className="glyphicon glyphicon-plus"/>
              </button>
              <button className="btn btn-default" type="button" style={{border:'1px solid #cccccc',borderLeft:'0px'}} onClick={()=>this.handleOpenSettingsModal()}>
                <span className="glyphicon glyphicon-cog"/>
              </button>
            </span>
        </div>
        { this.state.error !== "" ? <ErrorMessage message={this.state.error} closeError={this.clearError.bind(this)}/> : null}
      </form>
    )
  }
  handleOpenSettingsModal(){
    const {dispatch} = this.props;
    dispatch(openSettingsModal());
  }
  handleSubmit(e){
    e.preventDefault();
    const {dispatch} = this.props;
    if(this.state.value.length > 21)
      this.setState({error: "Subreddit must be 21 or less characters"});
    else if(!regex.test(this.state.value))
      this.setState({error: "Subreddit must contain only characters, numbers and _"});
    else{
        dispatch(fetchPostsIfNeeded(this.state.value));
        this.clearError();
    }
    this.setState({value: ""});

  }
  handleChange(event) {
    this.setState({value: event.target.value});
  }
  clearError(){
      this.setState({error: ""});
  }
  handleSettings(e){
    e.preventDefault();
  }
}

const RedditAdder = connect(null)(redditAdder);

export default RedditAdder;
