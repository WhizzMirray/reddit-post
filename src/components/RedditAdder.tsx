import React, { FormEvent } from "react";
import { connect } from "react-redux";

import { fetchPostsIfNeeded } from "../actions/redditActions";
import { openSettingsModal } from "../actions/utilActions";

import { DispatchFunc } from "../interfaces";
const regex = new RegExp("^[A-Za-z0-9][A-Za-z0-9_]{1,20}$"); // subreddit naming convention

const ErrorMessage = (message: any, closeError: any) => (
  <div
    className="alert alert-danger"
    role="alert"
    style={{ marginBottom: "0px", padding: "8px 15px" }}
  >
    <span
      className="glyphicon glyphicon-exclamation-sign"
      aria-hidden="true"
      style={{ paddingRight: "5px" }}
    />
    <span className="sr-only">Error:</span>
    <span className="pull-right clickable">
      <button
        className="error-close btn btn-default"
        type="button"
        onClick={closeError}
      >
        <span className="glyphicon glyphicon-remove" />
      </button>
    </span>
    {message}
  </div>
);

interface IProps {
  dispatch: DispatchFunc;
}

class RedditAdder extends React.Component<IProps> {
  public state = {
    error: "",
    value: ""
  };
  
  constructor(props: IProps) {
    super(props);
    this.handleOpenSettingsModal = this.handleOpenSettingsModal.bind(this);
    this.handleSubmit = this.handleSubmit.bind(this);
    this.handleChange = this.handleChange.bind(this);
    this.clearError = this.clearError.bind(this);
  }

  public render() {
    return (
      <form onSubmit={this.handleSubmit}>
        <div className="input-group">
          <input
            type="text"
            className="form-control"
            placeholder="Type the subreddit you want to add ex: 'worldnews'"
            onChange={this.handleChange}
            value={this.state.value}
          />
          <span className="input-group-btn">
            <button
              className="btn btn-default"
              type="submit"
              style={{ border: "1px solid #cccccc" }}
            >
              <span className="glyphicon glyphicon-plus" />
            </button>
            <button
              className="btn btn-default"
              type="button"
              style={{ border: "1px solid #cccccc", borderLeft: "0px" }}
              onClick={this.handleOpenSettingsModal}
            >
              <span className="glyphicon glyphicon-cog" />
            </button>
          </span>
        </div>
        {this.state.error !== "" ? (
          <ErrorMessage
            message={this.state.error}
            closeError={this.clearError}
          />
        ) : null}
      </form>
    );
  }

  public handleOpenSettingsModal() {
    const { dispatch } = this.props;
    dispatch(openSettingsModal());
  }

  public handleSubmit(e: FormEvent) {
    e.preventDefault();
    const { dispatch } = this.props;
    if (this.state.value.length > 21) {
      this.setState({ error: "Subreddit must be 21 or less characters" });
    } else if (!regex.test(this.state.value)) {
      this.setState({
        error: "Subreddit must contain only characters, numbers and _"
      });
    } else {
      dispatch(fetchPostsIfNeeded(this.state.value));
      this.clearError();
    }
    this.setState({ value: "" });
  }

  public handleChange(e: any){
    this.setState({ value: e.target.value });
  }

  public clearError() {
    this.setState({ error: "" });
  }
}

export default connect(null)(RedditAdder);
