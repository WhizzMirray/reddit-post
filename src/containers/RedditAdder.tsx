import * as React from "react";
import { connect } from "react-redux";
import { Dispatch } from "redux";


import { addSubreddit } from "../actions/listActions";
import { fetchPostsIfNeeded } from "../actions/redditActions";
import { openSettingsModal } from "../actions/utilActions";

import RedditAdderComponent from "../components/RedditAdder";

const regex = new RegExp("^[A-Za-z0-9][A-Za-z0-9_]{1,20}$"); // subreddit naming convention

export const ERROR_MESSAGES = {
  SUBREDDIT_MATCH : "Subreddit must contain only characters, numbers and _",
  SUBREDDIT_NAME_LENGTH : "Subreddit must be 21 or less characters",
}

interface IDispatchProps {
  handleOpenSettingsModal: () => void;
  handleSubmit: (subreddit: string) => void;
}

interface IState {
  error: string;
  value: string;
}

export type IProps = IDispatchProps; // Props = IStateProps & IDispatchProps & IOwnProps

export class RedditAdder extends React.Component<IProps, IState> {

  constructor(props: IProps) {
    super(props);
    this.handleSubmit = this.handleSubmit.bind(this);
    this.handleChange = this.handleChange.bind(this);
    this.clearError = this.clearError.bind(this);

    this.state = {
      error: "",
      value: ""
    };
  }

  public handleSubmit(e: React.FormEvent<HTMLFormElement>) {
    e.preventDefault();
    if (this.state.value.length > 21) {
      this.setState({ error: ERROR_MESSAGES.SUBREDDIT_NAME_LENGTH });
    } else if (!regex.test(this.state.value)) {
      this.setState({ error: ERROR_MESSAGES.SUBREDDIT_MATCH });
    } else {
      this.props.handleSubmit(this.state.value);
      this.clearError();
    }
    this.setState({ value: "" });
  }

  public handleChange = (e: any)  => {
    this.setState({ value: e.target.value });
  }

  public clearError() {
    this.setState({ error: "" });
  }

  public render() {
    const { handleOpenSettingsModal } = this.props;
    return (
      <RedditAdderComponent
        {...this.state}
        handleSubmit={this.handleSubmit}
        handleChange={this.handleChange}
        clearError={this.clearError}
        handleOpenSettingsModal={handleOpenSettingsModal}
      />
    );
  }
}
const mapDispatchToProps = (dispatch: Dispatch<any>): IDispatchProps => ({
  handleOpenSettingsModal: () => {
    dispatch(openSettingsModal());
  },
  handleSubmit: subreddit => {
    dispatch(addSubreddit(subreddit));
    dispatch(fetchPostsIfNeeded(subreddit));
  }
});
export default connect<null, IDispatchProps>(
  null,
  mapDispatchToProps
)(RedditAdder);
