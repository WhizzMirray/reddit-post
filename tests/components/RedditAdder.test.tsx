import { mount, shallow } from "enzyme";
import * as React from "react";

import RedditAdder, { IProps } from "../../src/components/RedditAdder";

const makeProps = (): IProps => ({
  clearError: jest.fn(),
  error: "",
  handleChange: jest.fn(),
  handleOpenSettingsModal: jest.fn(),
  handleSubmit: jest.fn(),
  value: ""
});

describe("<RedditAdder/>", () => {
  it("should render correctly", () => {
    const props = makeProps();
    const wrapper = shallow(<RedditAdder {...props} />);
    expect(wrapper).toMatchSnapshot();
  });

  it("should have <ErrorMessage> if error is not empty string", () => {
    const props = Object.assign(makeProps(), { error: "error" });
    const wrapper = shallow(<RedditAdder {...props} />);
    expect(wrapper).toMatchSnapshot();
  });

  it("should call handleOpenSettingsModal only once when the button is clicked", () => {
    const props = makeProps();
    const wrapper = shallow(<RedditAdder {...props} />);
    wrapper
      .find("button")
      .at(1)
      .simulate("click");

    expect(props.handleOpenSettingsModal).toHaveBeenCalledTimes(1);
  });

  it("should call handleChange when the input is changed", () => {
    const props = makeProps();
    const wrapper = shallow(<RedditAdder {...props} />);
    wrapper
      .find("input")
      .simulate("change", { target: { name: "", value: "hello" } });

    expect(props.handleChange).toBeCalledWith({
      target: { name: "", value: "hello" }
    });
  });

  it("should call handleSubmit when the button is clicked or the form submitted", () => {
    let props = makeProps();
    let wrapper = mount(<RedditAdder {...props} />);
    wrapper
      .find("button")
      .at(0)
      .simulate("submit");
    expect(props.handleSubmit).toHaveBeenCalledTimes(1);

    props = makeProps();
    wrapper = mount(<RedditAdder {...props} />);
    wrapper.find("form").simulate("submit");
    expect(props.handleSubmit).toHaveBeenCalledTimes(1);
  });
});
