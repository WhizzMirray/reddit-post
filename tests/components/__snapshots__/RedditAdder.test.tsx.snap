// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<RedditAdder/> should have <ErrorMessage> if error is not empty string 1`] = `
ShallowWrapper {
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <RedditAdder
    clearError={[MockFunction]}
    error="error"
    handleChange={[MockFunction]}
    handleOpenSettingsModal={[MockFunction]}
    handleSubmit={[MockFunction]}
    value=""
  />,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "host",
    "props": Object {
      "children": Array [
        <div
          className="input-group"
        >
          <input
            className="form-control"
            onChange={[MockFunction]}
            placeholder="Type the subreddit you want to add ex: 'worldnews'"
            type="text"
            value=""
          />
          <span
            className="input-group-btn"
          >
            <button
              className="btn btn-default"
              style={
                Object {
                  "border": "1px solid #cccccc",
                }
              }
              type="submit"
            >
              <span
                className="glyphicon glyphicon-plus"
              />
            </button>
            <button
              className="btn btn-default"
              onClick={[MockFunction]}
              style={
                Object {
                  "border": "1px solid #cccccc",
                  "borderLeft": "0px",
                }
              }
              type="button"
            >
              <span
                className="glyphicon glyphicon-cog"
              />
            </button>
          </span>
        </div>,
        <ErrorMessage
          clearError={[MockFunction]}
          message="error"
        />,
      ],
      "onSubmit": [MockFunction],
    },
    "ref": null,
    "rendered": Array [
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "children": Array [
            <input
              className="form-control"
              onChange={[MockFunction]}
              placeholder="Type the subreddit you want to add ex: 'worldnews'"
              type="text"
              value=""
            />,
            <span
              className="input-group-btn"
            >
              <button
                className="btn btn-default"
                style={
                  Object {
                    "border": "1px solid #cccccc",
                  }
                }
                type="submit"
              >
                <span
                  className="glyphicon glyphicon-plus"
                />
              </button>
              <button
                className="btn btn-default"
                onClick={[MockFunction]}
                style={
                  Object {
                    "border": "1px solid #cccccc",
                    "borderLeft": "0px",
                  }
                }
                type="button"
              >
                <span
                  className="glyphicon glyphicon-cog"
                />
              </button>
            </span>,
          ],
          "className": "input-group",
        },
        "ref": null,
        "rendered": Array [
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "host",
            "props": Object {
              "className": "form-control",
              "onChange": [MockFunction],
              "placeholder": "Type the subreddit you want to add ex: 'worldnews'",
              "type": "text",
              "value": "",
            },
            "ref": null,
            "rendered": null,
            "type": "input",
          },
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "host",
            "props": Object {
              "children": Array [
                <button
                  className="btn btn-default"
                  style={
                    Object {
                      "border": "1px solid #cccccc",
                    }
                  }
                  type="submit"
                >
                  <span
                    className="glyphicon glyphicon-plus"
                  />
                </button>,
                <button
                  className="btn btn-default"
                  onClick={[MockFunction]}
                  style={
                    Object {
                      "border": "1px solid #cccccc",
                      "borderLeft": "0px",
                    }
                  }
                  type="button"
                >
                  <span
                    className="glyphicon glyphicon-cog"
                  />
                </button>,
              ],
              "className": "input-group-btn",
            },
            "ref": null,
            "rendered": Array [
              Object {
                "instance": null,
                "key": undefined,
                "nodeType": "host",
                "props": Object {
                  "children": <span
                    className="glyphicon glyphicon-plus"
                  />,
                  "className": "btn btn-default",
                  "style": Object {
                    "border": "1px solid #cccccc",
                  },
                  "type": "submit",
                },
                "ref": null,
                "rendered": Object {
                  "instance": null,
                  "key": undefined,
                  "nodeType": "host",
                  "props": Object {
                    "className": "glyphicon glyphicon-plus",
                  },
                  "ref": null,
                  "rendered": null,
                  "type": "span",
                },
                "type": "button",
              },
              Object {
                "instance": null,
                "key": undefined,
                "nodeType": "host",
                "props": Object {
                  "children": <span
                    className="glyphicon glyphicon-cog"
                  />,
                  "className": "btn btn-default",
                  "onClick": [MockFunction],
                  "style": Object {
                    "border": "1px solid #cccccc",
                    "borderLeft": "0px",
                  },
                  "type": "button",
                },
                "ref": null,
                "rendered": Object {
                  "instance": null,
                  "key": undefined,
                  "nodeType": "host",
                  "props": Object {
                    "className": "glyphicon glyphicon-cog",
                  },
                  "ref": null,
                  "rendered": null,
                  "type": "span",
                },
                "type": "button",
              },
            ],
            "type": "span",
          },
        ],
        "type": "div",
      },
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "function",
        "props": Object {
          "clearError": [MockFunction],
          "message": "error",
        },
        "ref": null,
        "rendered": null,
        "type": [Function],
      },
    ],
    "type": "form",
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "host",
      "props": Object {
        "children": Array [
          <div
            className="input-group"
          >
            <input
              className="form-control"
              onChange={[MockFunction]}
              placeholder="Type the subreddit you want to add ex: 'worldnews'"
              type="text"
              value=""
            />
            <span
              className="input-group-btn"
            >
              <button
                className="btn btn-default"
                style={
                  Object {
                    "border": "1px solid #cccccc",
                  }
                }
                type="submit"
              >
                <span
                  className="glyphicon glyphicon-plus"
                />
              </button>
              <button
                className="btn btn-default"
                onClick={[MockFunction]}
                style={
                  Object {
                    "border": "1px solid #cccccc",
                    "borderLeft": "0px",
                  }
                }
                type="button"
              >
                <span
                  className="glyphicon glyphicon-cog"
                />
              </button>
            </span>
          </div>,
          <ErrorMessage
            clearError={[MockFunction]}
            message="error"
          />,
        ],
        "onSubmit": [MockFunction],
      },
      "ref": null,
      "rendered": Array [
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "children": Array [
              <input
                className="form-control"
                onChange={[MockFunction]}
                placeholder="Type the subreddit you want to add ex: 'worldnews'"
                type="text"
                value=""
              />,
              <span
                className="input-group-btn"
              >
                <button
                  className="btn btn-default"
                  style={
                    Object {
                      "border": "1px solid #cccccc",
                    }
                  }
                  type="submit"
                >
                  <span
                    className="glyphicon glyphicon-plus"
                  />
                </button>
                <button
                  className="btn btn-default"
                  onClick={[MockFunction]}
                  style={
                    Object {
                      "border": "1px solid #cccccc",
                      "borderLeft": "0px",
                    }
                  }
                  type="button"
                >
                  <span
                    className="glyphicon glyphicon-cog"
                  />
                </button>
              </span>,
            ],
            "className": "input-group",
          },
          "ref": null,
          "rendered": Array [
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "host",
              "props": Object {
                "className": "form-control",
                "onChange": [MockFunction],
                "placeholder": "Type the subreddit you want to add ex: 'worldnews'",
                "type": "text",
                "value": "",
              },
              "ref": null,
              "rendered": null,
              "type": "input",
            },
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "host",
              "props": Object {
                "children": Array [
                  <button
                    className="btn btn-default"
                    style={
                      Object {
                        "border": "1px solid #cccccc",
                      }
                    }
                    type="submit"
                  >
                    <span
                      className="glyphicon glyphicon-plus"
                    />
                  </button>,
                  <button
                    className="btn btn-default"
                    onClick={[MockFunction]}
                    style={
                      Object {
                        "border": "1px solid #cccccc",
                        "borderLeft": "0px",
                      }
                    }
                    type="button"
                  >
                    <span
                      className="glyphicon glyphicon-cog"
                    />
                  </button>,
                ],
                "className": "input-group-btn",
              },
              "ref": null,
              "rendered": Array [
                Object {
                  "instance": null,
                  "key": undefined,
                  "nodeType": "host",
                  "props": Object {
                    "children": <span
                      className="glyphicon glyphicon-plus"
                    />,
                    "className": "btn btn-default",
                    "style": Object {
                      "border": "1px solid #cccccc",
                    },
                    "type": "submit",
                  },
                  "ref": null,
                  "rendered": Object {
                    "instance": null,
                    "key": undefined,
                    "nodeType": "host",
                    "props": Object {
                      "className": "glyphicon glyphicon-plus",
                    },
                    "ref": null,
                    "rendered": null,
                    "type": "span",
                  },
                  "type": "button",
                },
                Object {
                  "instance": null,
                  "key": undefined,
                  "nodeType": "host",
                  "props": Object {
                    "children": <span
                      className="glyphicon glyphicon-cog"
                    />,
                    "className": "btn btn-default",
                    "onClick": [MockFunction],
                    "style": Object {
                      "border": "1px solid #cccccc",
                      "borderLeft": "0px",
                    },
                    "type": "button",
                  },
                  "ref": null,
                  "rendered": Object {
                    "instance": null,
                    "key": undefined,
                    "nodeType": "host",
                    "props": Object {
                      "className": "glyphicon glyphicon-cog",
                    },
                    "ref": null,
                    "rendered": null,
                    "type": "span",
                  },
                  "type": "button",
                },
              ],
              "type": "span",
            },
          ],
          "type": "div",
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "function",
          "props": Object {
            "clearError": [MockFunction],
            "message": "error",
          },
          "ref": null,
          "rendered": null,
          "type": [Function],
        },
      ],
      "type": "form",
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactFifteenAdapter {
      "options": Object {
        "lifecycles": Object {
          "componentDidUpdate": Object {
            "prevContext": true,
          },
        },
        "supportPrevContextArgumentOfComponentDidUpdate": true,
      },
    },
    "attachTo": undefined,
    "hydrateIn": undefined,
  },
}
`;

exports[`<RedditAdder/> should render correctly 1`] = `
ShallowWrapper {
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <RedditAdder
    clearError={[MockFunction]}
    error=""
    handleChange={[MockFunction]}
    handleOpenSettingsModal={[MockFunction]}
    handleSubmit={[MockFunction]}
    value=""
  />,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "host",
    "props": Object {
      "children": Array [
        <div
          className="input-group"
        >
          <input
            className="form-control"
            onChange={[MockFunction]}
            placeholder="Type the subreddit you want to add ex: 'worldnews'"
            type="text"
            value=""
          />
          <span
            className="input-group-btn"
          >
            <button
              className="btn btn-default"
              style={
                Object {
                  "border": "1px solid #cccccc",
                }
              }
              type="submit"
            >
              <span
                className="glyphicon glyphicon-plus"
              />
            </button>
            <button
              className="btn btn-default"
              onClick={[MockFunction]}
              style={
                Object {
                  "border": "1px solid #cccccc",
                  "borderLeft": "0px",
                }
              }
              type="button"
            >
              <span
                className="glyphicon glyphicon-cog"
              />
            </button>
          </span>
        </div>,
        null,
      ],
      "onSubmit": [MockFunction],
    },
    "ref": null,
    "rendered": Array [
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "children": Array [
            <input
              className="form-control"
              onChange={[MockFunction]}
              placeholder="Type the subreddit you want to add ex: 'worldnews'"
              type="text"
              value=""
            />,
            <span
              className="input-group-btn"
            >
              <button
                className="btn btn-default"
                style={
                  Object {
                    "border": "1px solid #cccccc",
                  }
                }
                type="submit"
              >
                <span
                  className="glyphicon glyphicon-plus"
                />
              </button>
              <button
                className="btn btn-default"
                onClick={[MockFunction]}
                style={
                  Object {
                    "border": "1px solid #cccccc",
                    "borderLeft": "0px",
                  }
                }
                type="button"
              >
                <span
                  className="glyphicon glyphicon-cog"
                />
              </button>
            </span>,
          ],
          "className": "input-group",
        },
        "ref": null,
        "rendered": Array [
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "host",
            "props": Object {
              "className": "form-control",
              "onChange": [MockFunction],
              "placeholder": "Type the subreddit you want to add ex: 'worldnews'",
              "type": "text",
              "value": "",
            },
            "ref": null,
            "rendered": null,
            "type": "input",
          },
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "host",
            "props": Object {
              "children": Array [
                <button
                  className="btn btn-default"
                  style={
                    Object {
                      "border": "1px solid #cccccc",
                    }
                  }
                  type="submit"
                >
                  <span
                    className="glyphicon glyphicon-plus"
                  />
                </button>,
                <button
                  className="btn btn-default"
                  onClick={[MockFunction]}
                  style={
                    Object {
                      "border": "1px solid #cccccc",
                      "borderLeft": "0px",
                    }
                  }
                  type="button"
                >
                  <span
                    className="glyphicon glyphicon-cog"
                  />
                </button>,
              ],
              "className": "input-group-btn",
            },
            "ref": null,
            "rendered": Array [
              Object {
                "instance": null,
                "key": undefined,
                "nodeType": "host",
                "props": Object {
                  "children": <span
                    className="glyphicon glyphicon-plus"
                  />,
                  "className": "btn btn-default",
                  "style": Object {
                    "border": "1px solid #cccccc",
                  },
                  "type": "submit",
                },
                "ref": null,
                "rendered": Object {
                  "instance": null,
                  "key": undefined,
                  "nodeType": "host",
                  "props": Object {
                    "className": "glyphicon glyphicon-plus",
                  },
                  "ref": null,
                  "rendered": null,
                  "type": "span",
                },
                "type": "button",
              },
              Object {
                "instance": null,
                "key": undefined,
                "nodeType": "host",
                "props": Object {
                  "children": <span
                    className="glyphicon glyphicon-cog"
                  />,
                  "className": "btn btn-default",
                  "onClick": [MockFunction],
                  "style": Object {
                    "border": "1px solid #cccccc",
                    "borderLeft": "0px",
                  },
                  "type": "button",
                },
                "ref": null,
                "rendered": Object {
                  "instance": null,
                  "key": undefined,
                  "nodeType": "host",
                  "props": Object {
                    "className": "glyphicon glyphicon-cog",
                  },
                  "ref": null,
                  "rendered": null,
                  "type": "span",
                },
                "type": "button",
              },
            ],
            "type": "span",
          },
        ],
        "type": "div",
      },
      null,
    ],
    "type": "form",
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "host",
      "props": Object {
        "children": Array [
          <div
            className="input-group"
          >
            <input
              className="form-control"
              onChange={[MockFunction]}
              placeholder="Type the subreddit you want to add ex: 'worldnews'"
              type="text"
              value=""
            />
            <span
              className="input-group-btn"
            >
              <button
                className="btn btn-default"
                style={
                  Object {
                    "border": "1px solid #cccccc",
                  }
                }
                type="submit"
              >
                <span
                  className="glyphicon glyphicon-plus"
                />
              </button>
              <button
                className="btn btn-default"
                onClick={[MockFunction]}
                style={
                  Object {
                    "border": "1px solid #cccccc",
                    "borderLeft": "0px",
                  }
                }
                type="button"
              >
                <span
                  className="glyphicon glyphicon-cog"
                />
              </button>
            </span>
          </div>,
          null,
        ],
        "onSubmit": [MockFunction],
      },
      "ref": null,
      "rendered": Array [
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "children": Array [
              <input
                className="form-control"
                onChange={[MockFunction]}
                placeholder="Type the subreddit you want to add ex: 'worldnews'"
                type="text"
                value=""
              />,
              <span
                className="input-group-btn"
              >
                <button
                  className="btn btn-default"
                  style={
                    Object {
                      "border": "1px solid #cccccc",
                    }
                  }
                  type="submit"
                >
                  <span
                    className="glyphicon glyphicon-plus"
                  />
                </button>
                <button
                  className="btn btn-default"
                  onClick={[MockFunction]}
                  style={
                    Object {
                      "border": "1px solid #cccccc",
                      "borderLeft": "0px",
                    }
                  }
                  type="button"
                >
                  <span
                    className="glyphicon glyphicon-cog"
                  />
                </button>
              </span>,
            ],
            "className": "input-group",
          },
          "ref": null,
          "rendered": Array [
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "host",
              "props": Object {
                "className": "form-control",
                "onChange": [MockFunction],
                "placeholder": "Type the subreddit you want to add ex: 'worldnews'",
                "type": "text",
                "value": "",
              },
              "ref": null,
              "rendered": null,
              "type": "input",
            },
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "host",
              "props": Object {
                "children": Array [
                  <button
                    className="btn btn-default"
                    style={
                      Object {
                        "border": "1px solid #cccccc",
                      }
                    }
                    type="submit"
                  >
                    <span
                      className="glyphicon glyphicon-plus"
                    />
                  </button>,
                  <button
                    className="btn btn-default"
                    onClick={[MockFunction]}
                    style={
                      Object {
                        "border": "1px solid #cccccc",
                        "borderLeft": "0px",
                      }
                    }
                    type="button"
                  >
                    <span
                      className="glyphicon glyphicon-cog"
                    />
                  </button>,
                ],
                "className": "input-group-btn",
              },
              "ref": null,
              "rendered": Array [
                Object {
                  "instance": null,
                  "key": undefined,
                  "nodeType": "host",
                  "props": Object {
                    "children": <span
                      className="glyphicon glyphicon-plus"
                    />,
                    "className": "btn btn-default",
                    "style": Object {
                      "border": "1px solid #cccccc",
                    },
                    "type": "submit",
                  },
                  "ref": null,
                  "rendered": Object {
                    "instance": null,
                    "key": undefined,
                    "nodeType": "host",
                    "props": Object {
                      "className": "glyphicon glyphicon-plus",
                    },
                    "ref": null,
                    "rendered": null,
                    "type": "span",
                  },
                  "type": "button",
                },
                Object {
                  "instance": null,
                  "key": undefined,
                  "nodeType": "host",
                  "props": Object {
                    "children": <span
                      className="glyphicon glyphicon-cog"
                    />,
                    "className": "btn btn-default",
                    "onClick": [MockFunction],
                    "style": Object {
                      "border": "1px solid #cccccc",
                      "borderLeft": "0px",
                    },
                    "type": "button",
                  },
                  "ref": null,
                  "rendered": Object {
                    "instance": null,
                    "key": undefined,
                    "nodeType": "host",
                    "props": Object {
                      "className": "glyphicon glyphicon-cog",
                    },
                    "ref": null,
                    "rendered": null,
                    "type": "span",
                  },
                  "type": "button",
                },
              ],
              "type": "span",
            },
          ],
          "type": "div",
        },
        null,
      ],
      "type": "form",
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactFifteenAdapter {
      "options": Object {
        "lifecycles": Object {
          "componentDidUpdate": Object {
            "prevContext": true,
          },
        },
        "supportPrevContextArgumentOfComponentDidUpdate": true,
      },
    },
    "attachTo": undefined,
    "hydrateIn": undefined,
  },
}
`;
